src/main/webapp/location/keyword.tsx

```diff
@@ -0,0 +1,302 @@
+import * as React from 'react'
+import {
+  geometry,
+  shapes,
+  coordinates as coordinateEditor,
+} from 'geospatialdraw'
+import { BasicEditorProps } from './geo-editor'
+import TextField from '@material-ui/core/TextField'
+import Autocomplete from '@material-ui/lab/Autocomplete'
+import CircularProgress from '@material-ui/core/CircularProgress'
+import Length from './length'
+import SpacedLinearContainer from '../spaced-linear-container'
+import { useApolloClient } from '@apollo/react-hooks'
+import { ApolloQueryResult } from 'apollo-client'
+import gql from 'graphql-tag'
+const { useApolloFallback } = require('../react-hooks')
+
+type QueryState = 'idle' | 'loading' | 'error'
+
+const SuggestionsQuery = gql`
+  query SuggestionsQuery($q: String!) {
+    suggestions(q: $q) {
+      id
+      name
+    }
+  }
+`
+
+const FeatureQuery = gql`
+  query FeatureQuery($id: String!) {
+    geofeature(id: $id) {
+      type
+      geometry
+      properties
+      id
+    }
+  }
+`
+
+type ContainerProps = BasicEditorProps & {
+  placeholder?: string
+  minimumInputLength?: number
+}
+
+type Props = ContainerProps & {
+  getSuggestions: (q: string) => Promise<Suggestion[]>
+  getGeoFeature: (id: string) => Promise<geometry.GeometryJSON>
+  loading: boolean
+  error: boolean
+}
+
+type Suggestion = {
+  id: string
+  name: string
+}
+
+type SelectedKeyword = {
+  keyword?: string
+  keywordId?: string
+}
+
+type KeywordGeoProperties = geometry.GeometryJSONProperties & SelectedKeyword
+
+const MINIMUM_INPUT_LENGTH = 2
+
+const Keyword: React.SFC<Props> = ({
+  value,
+  onChange,
+  placeholder = 'Pan to a region, country, or city',
+  minimumInputLength = MINIMUM_INPUT_LENGTH,
+  getSuggestions,
+  getGeoFeature,
+  loading,
+  error,
+}) => {
+  const {
+    id,
+    properties,
+    coordinates,
+    buffer,
+    bufferUnit,
+  } = coordinateEditor.geoToPolygonProps(value)
+  const {
+    keyword: initKeyword = '',
+    keywordId: initKeywordId,
+    color,
+  } = properties as KeywordGeoProperties
+  const hasSelection = initKeyword && initKeywordId ? true : false
+  const [input, setInput] = React.useState<string>(initKeyword)
+  const [open, setOpen] = React.useState(false)
+  const [suggestions, setSuggestions] = React.useState<Suggestion[]>([])
+  const [{ keyword, keywordId }, setSelectedKeyword] = React.useState<
+    SelectedKeyword
+  >({})
+  React.useEffect(
```

Comment:
I think this code could be simplified. I think we have too many effects happening.